nav2_loopback_sim/src/loopback_simulator.cpp

```diff
@@ -0,0 +1,227 @@
+  geometry_msgs::msg::Pose new_position;
+  new_position.position.x =
+    transformStamped.transform.translation.x + rotated_linear_velocity.x() * dt_;
+  new_position.position.y =
+    transformStamped.transform.translation.y + rotated_linear_velocity.y() * dt_;
+  new_position.position.z =
+    transformStamped.transform.translation.z + rotated_linear_velocity.z() * dt_;
+  new_position.orientation = tf2::toMsg(new_orientation);
+
+  // Publish the updated transform between odom and base_link
+  publishTransform(new_position, "odom", "base_link");
+}
+
+void LoopbackSimulator::initPoseCallback(
+  const
+  geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+  RCLCPP_INFO(this->get_logger(), "Received initial pose!");
+  init_pose_ = *msg;
+  init_pose_set_ = true;
+
+  if (!zero_odom_base_published_) {
+    // Publish first init pose between map and odom
+    publishTransform(init_pose_.pose.pose, "map", "odom");
+    geometry_msgs::msg::Pose zero_pose{};
+
+    // Publish identity transform between odom and base_footprint
+    publishTransform(zero_pose, "odom", "base_footprint");
+
+    relocalization_pose_ = init_pose_.pose.pose;
+    zero_odom_base_published_ = true;
+    return;
+  }
+
+  // Updating map->base_link transform based on set pose
+  geometry_msgs::msg::TransformStamped map_to_base_link_transform;
+  map_to_base_link_transform.header = init_pose_.header;
+  map_to_base_link_transform.header.frame_id = "map";
+  map_to_base_link_transform.child_frame_id = "base_link";
+  map_to_base_link_transform.transform.translation.x = init_pose_.pose.pose.position.x;
+  map_to_base_link_transform.transform.translation.y = init_pose_.pose.pose.position.y;
+  map_to_base_link_transform.transform.translation.z = init_pose_.pose.pose.position.z;
+  map_to_base_link_transform.transform.rotation = init_pose_.pose.pose.orientation;
+
+  geometry_msgs::msg::TransformStamped odom_to_base_link_transform;
+  try {
+    odom_to_base_link_transform = tf_buffer_->lookupTransform(
+      "odom",
+      "base_link",
+      tf2::TimePointZero);
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_ERROR(this->get_logger(), "%s", ex.what());
+    return;
+  }
+
+  // Calculating map->odom transform
+  geometry_msgs::msg::TransformStamped map_to_odom_transform;
+  tf2::doTransform(map_to_base_link_transform, map_to_odom_transform, odom_to_base_link_transform);
+
+  // Updating relocalization pose of odom frame
+  relocalization_pose_.position.x = map_to_odom_transform.transform.translation.x;
+  relocalization_pose_.position.y = map_to_odom_transform.transform.translation.y;
+  relocalization_pose_.position.z = map_to_odom_transform.transform.translation.z;
+  relocalization_pose_.orientation = map_to_odom_transform.transform.rotation;
+
+  // Publishing updated position of odom and base_link wrt map
+  publishTransform(relocalization_pose_, "map", "odom");
+  publishTransform(init_pose_.pose.pose, "map", "base_link");
```

Comment:
The transform it should be publishing is odom->base_link, the map->base_link should **never** be published. That breaks the TF tree. Its map->odom and odom->base_link. I think that largely invalidates the code above.

- Given the initial pose (map->base_link)
- Find the current odom->base_link transform (or just have it stored from the previous iteration)
- Back out the map->odom that keeps odom->base_link constant 
- Publish that map->odom